File: apps/ridi-mobile/modules/ridi-router/android/src/main/java/bike/ridi/modules/router/RidiRouterModule.kt
package bike.ridi.modules.router

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject
import uniffi.ridi_router_mobile.generateRouteJson as generateRouteJsonRust

class RidiRouterModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RidiRouter")

    AsyncFunction("generateRouteJson") { tilesDir: String, requestJson: String ->
      try {
        generateRouteJsonRust(tilesDir, requestJson)
      } catch (throwable: Throwable) {
        nativeBindingFailedEnvelope(throwable)
      }
    }
  }

  private fun nativeBindingFailedEnvelope(throwable: Throwable): String {
    return "{" +
      "\"ok\":false," +
      "\"error\":{" +
      "\"code\":\"native_binding_failed\"," +
      "\"message\":\"Native routing call failed\"," +
      "\"details\":" + JSONObject.quote(throwable.toString()) +
      "}" +
      "}"
  }
}

File: apps/ridi-mobile/modules/ridi-router/ios/RidiRouterModule.swift
import ExpoModulesCore

public class RidiRouterModule: Module {
  public func definition() -> ModuleDefinition {
    Name("RidiRouter")

    AsyncFunction("generateRouteJson") { (tilesDir: String, requestJson: String) -> String in
      do {
        return try generateRouteJson(tilesDir: tilesDir, requestJson: requestJson)
      } catch {
        return Self.nativeBindingFailedEnvelope(error)
      }
    }
  }

  private static func nativeBindingFailedEnvelope(_ error: Error) -> String {
    let details = String(describing: error)
    let envelope: [String: Any] = [
      "ok": false,
      "error": [
        "code": "native_binding_failed",
        "message": "Native routing call failed",
        "details": details,
      ],
    ]

    guard JSONSerialization.isValidJSONObject(envelope),
      let data = try? JSONSerialization.data(withJSONObject: envelope),
      let json = String(data: data, encoding: .utf8)
    else {
      return "{\"ok\":false,\"error\":{\"code\":\"unknown\",\"message\":\"Native routing call failed\"}}"
    }

    return json
  }
}
